src/olympia/users/admin.py

```diff
@@ -53,6 +63,101 @@ class UserAdmin(admin.ModelAdmin):
+                wrap(self.delete_picture_view),
+                name='users_userprofile_delete_picture')
+        ]
+        return custom_urlpatterns + urlpatterns
+
+    def get_actions(self, request):
+        actions = super(UserAdmin, self).get_actions(request)
+        if not acl.action_allowed(request, amo.permissions.RATINGS_MODERATE):
+            # You need Ratings:Edit to be able to ban users.
+            actions.pop('user_ban', None)
+        return actions
+
+    def change_view(self, request, object_id, form_url='', extra_context=None):
+        extra_context = extra_context or {}
+        extra_context['can_ban'] = acl.action_allowed(
+            request, amo.permissions.RATINGS_MODERATE)
+        return super(UserAdmin, self).change_view(
+            request, object_id, form_url, extra_context=extra_context,
+        )
+
+    def delete_model(self, request, obj):
+        # Deleting a user through the admin also deletes related content
+        # produced by that user.
+        ActivityLog.create(amo.LOG.ADMIN_USER_ANONYMIZED, obj)
+        obj.delete_related_content()
+        obj.delete()
+
+    def save_model(self, request, obj, form, change):
+        # what changes? details=self.changes()
+        changes = {k: (form.initial.get(k), form.cleaned_data.get(k))
+                   for k in form.changed_data}
+        ActivityLog.create(amo.LOG.ADMIN_USER_EDITED, obj, details=changes)
+        obj.save()
+
+    def ban_view(self, request, object_id, extra_context=None):
+        if request.method != 'POST':
+            return HttpResponseNotAllowed(['POST'])
+
+        obj = self.get_object(request, unquote(object_id))
+        if obj is None:
+            raise Http404()
+
+        if not acl.action_allowed(request, amo.permissions.RATINGS_MODERATE):
+            return HttpResponseForbidden()
+
+        ActivityLog.create(amo.LOG.ADMIN_USER_BANNED, obj)
+        obj.ban_and_delete_related_content()
+        self.message_user(
+            request, ugettext('The user "%(user)s" has been banned.' %
+                              {'user': force_text(obj)}))
+        return HttpResponseRedirect('../')
+
+    def delete_picture_view(self, request, object_id, extra_context=None):
+        if request.method != 'POST':
+            return HttpResponseNotAllowed(['POST'])
+
+        obj = self.get_object(request, unquote(object_id))
+        if obj is None:
+            raise Http404()
+
+        if not acl.action_allowed(request, amo.permissions.USERS_EDIT):
+            return HttpResponseForbidden()
+
+        ActivityLog.create(amo.LOG.ADMIN_USER_PICTURE_DELETED, obj)
+        obj.delete_picture()
+        self.message_user(
+            request, ugettext(
+                'The picture belonging to user "%(user)s" has been deleted.' %
+                {'user': force_text(obj)}))
+        return HttpResponseRedirect('../')
+
+    def ban_action(self, request, qs):
+        users = []
```

Comment:
should there be an `acl.action_allowed(...)` check here like there is for `ban_view`?

Reply:
We could, but in theory django should never let you call this, because it gets filtered by `get_actions()` before that. I'll double-check.